glommio/src/io/bulk_io.rs

```diff
@@ -49,3 +49,52 @@ impl<U: Copy + Unpin> Stream for OrderedBulkIo<U> {
         (self.iovs.len(), Some(self.iovs.len()))
     }
 }
+
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct ReadManyArgs {
+    pub(crate) user_read: (u64, usize),
+    pub(crate) system_read: (u64, usize),
+}
+
+#[derive(Debug)]
+pub struct ReadManyResult {
+    pub(crate) inner: OrderedBulkIo<ReadManyArgs>,
+    pub(crate) current_result: ReadResult,
+}
+
+impl Stream for ReadManyResult {
+    type Item = super::Result<(u64, ReadResult)>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let next = match Pin::new(&mut self.inner).poll_next(cx) {
+            Poll::Ready(next) => next,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        match next {
+            None => Poll::Ready(None),
+            Some((source, args)) => {
+                if let Some(mut source) = source {
+                    let read_size =
+                        enhanced_try!(source.take_result().unwrap(), "Reading", self.inner.file)?;
+                    let mut buffer = source.extract_dma_buffer();
+                    buffer.trim_to_size(std::cmp::min(read_size, args.system_read.1));
+                    self.current_result = ReadResult::from_whole_buffer(buffer);
+                }
+                Poll::Ready(Some(Ok((
```

Comment:
I am having a really hard time understanding this:
every time you poll, if there is a Source, you process that `Source`. And then you return a single slice from it.

How can you return multiple slices from a single `Source` if you never test for `current_result` in your poll function?

Reply:
If the inner stream yields `(_, Some(Source))` then I create a `ReadResult` from it and return that (trimmed according to the current iovec). If the inner stream yields `(_, None)`, then I reuse the last `ReadResult`. This way I let go of the source as soon as possible but I keep the underlying buffer for as long as necessary.

Reply:
ok. Since this is properly unit tested, I'm essentially ok with whatever is fast.
(And what determines what is fast is data, not my personal taste anyway)